### ISSUE TITLE
Attempt reasonable type conversions for query arguments

### ISSUE THREAD

**Issue by gadomski** [stac-utils/stac-fastapi#202]
Original issue: https://github.com/stac-utils/pystac-client/issues/80.

`pystac-client` allows the user to specify a query via the command line, e.g.:

```bash
$ stac-client search --query "gsd=30" ...
```

Becuase `pystac-client` doesn't have any typing information, it can't know what type to convert that `30` to before sending the query, so it just sends it as a string. This leads to an always-`False` query on the `stac-fastapi` side. The `pystac-client` user could work around the issue by specifying the query dictionary exactly (`{"gsd": {"eq": 30}}`) but that's not as command-line friendly.

Since `stac-fastapi` has typing information, would it be reasonable to attempt to convert that incoming `30` to the correct type performing the query?

Per https://github.com/stac-utils/pystac-client/issues/80#issuecomment-891353826, `stac-server` does allow strings for this type of query.

**Comment by geospatial-jeff** [comment]
It's possible but it could be difficult.  POST requests use the `stac_pydantic.api.Search` model (https://github.com/stac-utils/stac-pydantic/blob/master/stac_pydantic/api/search.py#L36).  In order to do this we'd have to define a stricter type for `Search.query` than what is currently provided - `Optional[Dict[str, Dict[Operator, Any]]]`.

We could do this by defining more pydantic models for the `Search.query` object, but we would have to do this explicitely for each field the user wants to search on, something like this:

```python
class EqualsInteger(BaseModel):
    eq: int

class StacQuery(BaseModel):
    gsd: EqualsInteger

query = StacQuery(gsd=EqualsInteger(eq="23"))
assert isinstance(query.gsd.eq, int)
```

Because STAC is extendible this pattern isn't that sustainable.  Best case scenario is we type out the queries that most people will do (gsd, cloud_cover etc.) and allow other queries to be passed through the model (ex. if someone has their own extension they want to query against).  These extra queries wouldn't be type checked though.

GET requests functionally work a little bit differently in terms of how fastapi validates them but its basically the same scenario.

**Comment by gadomski** [comment]
> Best case scenario is we type out the queries that most people will do (gsd, cloud_cover etc.) and allow other queries to be passed through the model (ex. if someone has their own extension they want to query against).

Yeah, unfortunate that there has to be boilerplate for every attribute we'd want to support in this way. Looking through https://github.com/radiantearth/stac-spec/blob/master/item-spec/common-metadata.md, `gsd` is the only non-string non-datetime attribute on that list, so it could be considered a special case worth supporting. Agreed though that `cloud_cover` and stuff like that would be helpful too, though where does it stop?

**Comment by geospatial-jeff** [comment]
>  though where does it stop?

This.  it's a very slippery slope.

What would be helpful in the spec is to provide a way for servers which implement the query extension to specify which fields are queryable.  The filter extension has the notion of queryables (https://github.com/radiantearth/stac-api-spec/tree/master/fragments/filter#queryables).  Type information is exposed through JSON schema, if the server can publish ahead of time which fields the user can query on it wouldn't be too difficult to programmatically create the stac pydantic `Search` model on application startup.

**Comment by gadomski** [comment]
> What would be helpful in the spec is to provide a way for servers which implement the query extension to specify which fields are queryable.

Right, and without knowing too much backstory myself, it looks like the query extension is [deprecated](https://github.com/radiantearth/stac-api-spec/blob/master/extensions.md#list-of-community-extensions) so it seems questionable how much effort is justified.

Does the query engine give back any type-mismatch information from a request that could be used to inform the user of what went wrong?

**Comment by geospatial-jeff** [comment]
> Right, and without knowing too much backstory myself, it looks like the query extension is deprecated so it seems questionable how much effort is justified.

Deprecated in favor of [filter extension](https://github.com/radiantearth/stac-api-spec/tree/master/fragments/filter) which does have a notion of queryables.

> Does the query engine give back any type-mismatch information from a request that could be used to inform the user of what went wrong?

Honesty I'm not sure.  It wouldn't be caught by pydantic's request validation, so this would probably happen in the backends which means you probably can't rely on this type-mismatch information to be reliable.
